Sources/PackageModel/PackageIdentity.swift

```diff
@@ -0,0 +1,196 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Foundation
+
+import TSCBasic
+import TSCUtility
+
+public struct PackageIdentity: LosslessStringConvertible {
+    public let description: String
+    public let computedName: String
+
+    public init(_ string: String) {
+        var string = string.precomposedStringWithCanonicalMapping
+
+        var detectedScheme: Scheme?
+        for scheme in Scheme.allCases {
+            if string.removePrefixIfPresent("\(scheme):") {
+                detectedScheme = scheme
+                string.removePrefixIfPresent("//")
+                break
+            }
+        }
+
+        string.removePortComponentIfPresent()
+
+        if string.removeUserComponentIfPresent() || detectedScheme != .ssh {
+            string.replaceFirstOccurenceIfPresent(of: ":", with: "/")
+        }
+
+        string.removeFragmentComponentIfPresent()
+        string.removeQueryComponentIfPresent()
+
+        var components = string.split(omittingEmptySubsequences: true, whereSeparator: isSeparator)
+
+        var lastPathComponent = components.popLast() ?? ""
+        lastPathComponent.removeSuffixIfPresent(".git")
+        components.append(lastPathComponent)
+
+        self.description = components.joined(separator: "/")
+        self.computedName = String(lastPathComponent)
+    }
+}
+
+extension PackageIdentity: Equatable {
+    public static func == (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.description == rhs.description
+    }
+}
+
+extension PackageIdentity: Comparable {
+    public static func < (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.description < rhs.description
+    }
+}
+
+extension PackageIdentity: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(self.description)
+    }
+}
+
+extension PackageIdentity: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let description = try container.decode(String.self)
+        self.init(description)
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.description)
+    }
+}
+
+extension PackageIdentity: JSONMappable, JSONSerializable {
```

Comment:
I assume this is not Codable because of upstream/downstream code that does not use Codable?

Reply:
Yes, correct. I have no strong feelings either way as far as keeping existing legacy `JSON` conformance or migrating to `Codable`, so let me know what you'd like to do here.

Reply:
We should migrate to Codable.  The only reason for hand rolling JSON in SwiftPM/TSC was that the functionality didn't yet exist five years ago.